_source/_posts/2022-01-19-react-storybook.md

```diff
@@ -0,0 +1,688 @@
+  onLogin: PropTypes.func.isRequired,
+  onLogout: PropTypes.func.isRequired,
+  onCreateAccount: PropTypes.func.isRequired,
+};
+
+Header.defaultProps = {
+  user: null,
+};
+```
+
+I have modified the header component to show the correct application name and allow some structured user data to be passed in. In the story for the header, in file `src/stories/Header.stories.jsx`, modify the arguments passed to the `LoggedIn` story to reflect this change.
+
+```js
+LoggedIn.args = {
+  user: {
+    given_name: "Username"
+  },
+};
+```
+
+Now, open `src/stories/Page.jsx` and modify its contents to match the code below.
+
+```js
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Header } from './Header';
+import './page.css';
+import { Tabs } from './Components';
+import { Converter } from './Converter';
+
+export const Page = ({useAuth}) => {
+  const [user, login, logout] = useAuth();
+  return <article>
+    <Header user={user} onLogin={login} onLogout={logout} />
+    <section>
+      <Tabs>
+        <Converter measure="length" label="Length" key="length"></Converter>
+        <Converter measure="mass" label="Mass" key="mass"></Converter>
+        <Converter measure="volume" label="Volume" key="volume"></Converter>
+      </Tabs>
+    </section>
+  </article>;
+}
+
+Page.propTypes = {
+  useAuth: PropTypes.func.isRequired
+};
+
+Page.defaultProps = {
+};
+```
+
+This component displays the application page, including the header and a tabbed container that allows switching between `Converter` components configured to convert different measures. The page needs a `useAuth` hook passed in that returns the user information and callbacks to log the user in or out. In the stories for the page, in `src/stories/Page.stories.jsx`, you need to create a mock function that supplies fake user data. Edit the contents of this file to look like the following code.
+
+```js
+import React from 'react';
+import { Page } from './Page';
+
+export default {
+  title: 'Pages/Page',
+  component: Page,
+};
+
+const mockUseAuth = (loggedIn) => () => [
+  loggedIn ? {given_name: "Username"} : undefined, 
+  () => {}, 
+  () => {}
+];
+
+const Template = (args) => <Page useAuth={mockUseAuth(true)} {...args}/>;
+
+export const LoggedIn = Template.bind({});
+LoggedIn.args = {
+  useAuth: mockUseAuth(true),
+};
+
+LoggedIn.parameters = {
+  controls: { hideNoControlsWarning: true },
+};
+
+export const LoggedOut = Template.bind({});
+LoggedOut.args = {
+  useAuth: mockUseAuth(false),
+};
+
+LoggedOut.parameters = {
+  controls: { hideNoControlsWarning: true },
+};
+```
+
+Note how `mockUseAuth` uses currying to return a function that can be used as the `useAuth` hook in the `Page` component. You can now use Storybook again to test the `Converter` component and the full application page. If it's not still running, run `npm run storybook` again. You can navigate to **Pages -> Page** in the left sidebar, and you should see something like the image below.
```

Comment:
Would it make sense to 'define' currying by providing a link? https://en.wikipedia.org/wiki/Currying ? It was not familiar to me.

Reply:
Yep, will do. But I think this one (unless there's an objection) https://javascript.info/currying-partials